design-documents/graph-ql/coverage/returns.graphqls

```diff
@@ -0,0 +1,210 @@
+    entered_custom_attributes: [EnteredCustomAttributeInput!] @doc(description: "Values of return item attributes defined by the merchant, e.g. file or text attributes.")
+}
+
+input EnteredCustomAttributeInput {
+    id: ID!
+    value: String!
+}
+
+type RequestReturnOutput {
+    return: Return
+    returns(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. Defaults to 20."),
+        currentPage: Int = 1 @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): Returns @doc(description: "Information about the customer returns.")
+}
+
+input AddReturnCommentInput {
+    return_id: ID!
+    comment_text: String!
+}
+
+type AddReturnCommentOutput {
+    return: Return
+}
+
+input AddReturnTrackingInput {
+    return_id: ID!
+    carrier_id: ID!
+    tracking_number: String!
+}
+
+type AddReturnTrackingOutput {
+    return: Return
+}
+
+input RemoveReturnTrackingInput {
+    return_shipping_tracking_id: ID!
+}
+
+type RemoveReturnTrackingOutput {
+    return: Return
+}
+
+enum CustomAttributesPageEnum {
+    RETURN_ITEM_EDIT_FORM
+    RETURN_ITEMS_LISTING
+    # See https://github.com/magento/architecture/blob/master/design-documents/graph-ql/framework/attributes-metadata.md
+    # PRODUCTS_COMPARE
+    # PRODUCTS_LISTING
+    # ADVANCED_CATALOG_SEARCH
+}
+
+type Customer {
+    returns(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. Defaults to 20."),
+        currentPage: Int = 1 @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): Returns @doc(description: "Information about the customer returns.")
+    return(id: ID!): Return @doc(description: "Get customer return details by its ID.")
+}
+
+type CustomerOrder {
+    returns(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. Defaults to 20."),
+        currentPage: Int = 1 @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): Returns @doc(description: "Returns associated with this order.")
+    items_eligible_for_return: [OrderItemInterface] @doc(description: "A list of order items eligible for return.")
+}
+
+type OrderItemInterface {
+    eligible_for_return: Boolean @doc(description: "Indicates whether the order item is eligible for return.")
+}
+
+type Returns {
+    items: [Return] @doc(description: "List of returns")
+    page_info: SearchResultPageInfo @doc(description: "Pagination metadata")
+    total_count: Int @doc(description: "Total count of customer returns")
+}
+
+type Return @doc(description: "Customer return") {
+    id: ID! @doc(description: "Return number")
+    order_id: ID! @doc(description: "ID of the order based on which the return was created.")
```

Comment:
Should probably be the order object - clients can then grab the `ID` from it if they want. Otherwise we have another missing edge in the graph

Reply:
I think it definitely makes sense from developer usage perspective but it will have performance impact on the current implementation as order entity overloaded and very "heavy".

Reply:
> it will have performance impact on the current implementation as order entity overloaded and very "heavy".

It'll be more of a performance impact if the client has to make multiple requests, and spin up a new request thread + run magento bootstrap for each one.

There are ways we can address performance impacts. If the only field requested is the `id`, you can design the resolver to only grab the ID.

Reply:
Usually, for RMA we just show only order number as a reference to original order.

> There are ways we can address performance impacts. If the only field requested is the id, you can design the resolver to only grab the ID.

I'm not sure if we have something similar right now, as far as I know, we extract all entity fields from the database. In the case with an order entity, it's important to take into account because it's a "heavy" entity with a lot of relations.

Reply:
> Usually, for RMA we just show only order number as a reference to original order.

Right, agree that's how it works today in Luma. As we move to becoming a headless platform with a 1st class API, though, the UX requirements won't be set by us anymore, so it'll be our job to make data reachable in the graph.

> I'm not sure if we have something similar right now, as far as I know, we extract all entity fields from the database. In the case with an order entity, it's important to take into account because it's a "heavy" entity with a lot of relations.

Code won't need to load the entire order entity from the database, though; only if that data is requested by the client. If our resolvers are written to load _all_ data, regardless of whether it was selected by the client, sounds more like technical debt we can improve upon.

Reply:
> Code won't need to load the entire order entity from the database, though; only if that data is requested by the client. If our resolvers are written to load all data, regardless of whether it was selected by the client, sounds more like technical debt we can improve upon.

I completely agree that the code shouldn't load the entire entity. My concern that in the near feature will have only current business domain implementation which has issues with performance and doesn't have such partial loading.
I think we always can extend the schema and add a reference to the order entity instead of the order number.